consensus/event_loop.go

```diff
@@ -0,0 +1,854 @@
+		// Signal the server that we can construct a timeout QC for the current view
+		fc.hasConstructedQCInCurrentView = true
+		fc.Events <- timeoutQCEvent
+		return
+	}
+}
+
+// tryConstructVoteQCInCurrentView is a helper function that attempts to construct a QC for the tip block
+// so that it can be proposed in a block in the current view. The function internally performs all view and vote
+// validations to ensure that the resulting QC is valid. If a QC can be constructed, the function returns
+// the signers list and aggregate signature that can be used to construct the QC.
+//
+// This function must be called while holding the event loop's lock.
+func (fc *FastHotStuffEventLoop) tryConstructVoteQCInCurrentView() *FastHotStuffEvent {
+	// If currentView != tipBlock.View + 1, then we have timed out at some point, and can no longer
+	// construct a block with a QC of votes for the tip block.
+	tipBlock := fc.tip.block
+	if fc.currentView != tipBlock.GetView()+1 {
+		return nil
+	}
+
+	// Fetch the validator set at the tip.
+	validatorSet := fc.tip.validatorSet
+
+	// Compute the chain tip's signature payload.
+	voteSignaturePayload := GetVoteSignaturePayload(tipBlock.GetView(), tipBlock.GetBlockHash())
+
+	// Fetch the validator votes for the tip block.
+	votesByValidator := fc.votesSeen[voteSignaturePayload]
+
+	// Compute the total stake and total stake with votes
+	totalStake := uint256.NewInt()
+	totalVotingStake := uint256.NewInt()
+
+	// Track the signatures and signers list for the chain tip
+	signersList := bitset.NewBitset()
+	signatures := []*bls.Signature{}
+
+	// Iterate through the entire validator set and check if each one has voted for the tip block. Track
+	// all voters and their stakes.
+	for ii, validator := range validatorSet {
+		totalStake = uint256.NewInt().Add(totalStake, validator.GetStakeAmount())
+
+		// Skip the validator if it hasn't voted for the the block
+		vote, hasVoted := votesByValidator[validator.GetPublicKey().ToString()]
+		if !hasVoted {
+			continue
+		}
+
+		// Verify the vote signature
+		if !isValidSignatureSinglePublicKey(vote.GetPublicKey(), vote.GetSignature(), voteSignaturePayload[:]) {
+			continue
+		}
+
+		// Track the vote's signature, stake, and place in the validator set
+		totalVotingStake = uint256.NewInt().Add(totalVotingStake, validator.GetStakeAmount())
+		signersList.Set(ii, true)
+		signatures = append(signatures, vote.GetSignature())
+	}
+
+	// If we don't have a super-majority vote for the chain tip, then we can't build a QC.
+	if !isSuperMajorityStake(totalVotingStake, totalStake) {
+		return nil
+	}
+
+	// If we reach this point, then we have enough signatures to build a QC for the tip block. Try to
+	// aggregate the signatures. This should never fail.
+	aggregateSignature, err := bls.AggregateSignatures(signatures)
+	if err != nil {
+		return nil
+	}
+
+	// Happy path. Construct the QC and return as an event to signal to the server.
+	return &FastHotStuffEvent{
```

Comment:
Is there any way the view or block tip could advance between the start of this function and this section? I think no because 1) The caller of this function grabs a lock and 2) AdvanceView* grabs a lock always. But just double-checking.

Reply:
Your reasoning is correct. Neither the view nor the block tip can shift underneath us here. This function's usage is thread safe.